### PR TITLE
Composed provider

### PR DESCRIPTION
**Pull request by Velenir**
Experiment to compose providers with `json-rpc-engine`
~Hardcoded to work with MMask and on rinkeby
Uses a separate WebSocket provider for subscriptions
Seems to work, but probably breaks in a thousand hidden places~

MMask subscriptions work if we reemit a bunch of events ('data' event is enough, but just in case)
Same for WalletConnect provider

Possible application -- `tx.data` earmarking, `eth_gasPrice` delegation to a gas station

From brief testing, everything seems to work.
I would like to remake gas price fetching and `tx.data` earmarking into middlewares

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr1008--dexreact.review.gnosisdev.com

File: src/api/wallet/composeProvider.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  const walletMware = providerAsMiddleware(provider)
+
+  // fork middleware
+  // req is handled by either MMask prov or WS prov
+  engine.push((req, res, next, end) => {
+    console.log('REQ', req)
+    if (req.method === 'eth_subscribe' && req.params[0] === 'newHeads') {
+      console.log('DELEGATING TO WS', req.params[0])
+      // an interesting idea would be to use MMask for eth_sign|sendTransaction
+      // ans WS for everything else
```
Comment (alfetopito) [inline src/api/wallet/composeProvider.ts:76]: What's the advantage?
Isn't it better for us if we delegate what we can to user's MM provider, using less of our own Infura bandwidth?

File: src/api/wallet/composeProvider.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      end()
+    })
+  }
+}
+
+const WS_URL = 'wss://rinkeby.infura.io/ws/v3/607a7dfcb1ad4a0b83152e30ce20cfc5'
+const createWSProvider = (url: string): WebsocketProvider => new Web3.providers.WebsocketProvider(url)
+
+export const composeProviderWithWS = (provider: Provider): Provider => {
+  const engine = new (RpcEngine as any)() as JsonRpcEngine
```
[the remark this answers is not in the document]
Comment (Velenir) [reply src/api/wallet/composeProvider.ts:34]: Their types are incomplete and we can't extend without overriding them entirely

File: src/api/wallet/composeProvider.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+  const WSProvider = createWSProvider(WS_URL)
+  // WSProvider.on('connect', () => console.log('CPROV::WS connect'))
+  // // @ts-ignore
+  // WSProvider.on('reconnect', (attempts: number) => console.log('CPROV::WS reconnect', attempts))
+  // // @ts-ignore
+  // WSProvider.on('close', (ev: any): void => console.log('CPROV::WS close', ev))
+  // // @ts-ignore
+  // WSProvider.on('error', (error: Error) => console.log('CPROV::WS error', error))
+
+  const WSMware = providerAsMiddleware(WSProvider)
```
[the remark this answers is not in the document]
Comment (Velenir) [reply src/api/wallet/composeProvider.ts:59]: MWare stands for Malware here

File: src/api/wallet/composeProvider.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+  engine.push((req, res, next, end) => {
+    console.log('CPROV::REQ', req)
+    if (req.method === 'eth_subscribe' && req.params[0] === 'newHeads') {
+      console.log('CPROV::DELEGATING TO WS', req.params[0])
+      // an interesting idea would be to use MMask for eth_sign|sendTransaction
+      // ans WS for everything else
+      WSMware(req, res, next, end)
+    } else {
+      // MMask provider can't handle eth_subscribe
+      walletMware(req, res, next, end)
```
[the remark this answers is not in the document]
Comment (Velenir) [reply src/api/wallet/composeProvider.ts:74]: That was just me throwing staff at it and seeing what works :grin: 

Turns out can use wallet for subscriptions too, but need to reemit events

File: src/api/wallet/composeProvider.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/ban-ts-ignore */
+import RpcEngine, { JsonRpcEngine, JsonRpcMiddleware, JsonRpcResponse, JsonRpcError } from 'json-rpc-engine'
+// import providerAsMiddleware from 'eth-json-rpc-middleware/providerAsMiddleware'
+import providerFromEngine from 'eth-json-rpc-middleware/providerFromEngine'
```
[the remark this answers is not in the document]
Comment (Velenir) [reply src/api/wallet/composeProvider.ts:5]: 6.72 Kb and less than half of that when gzipped
Drop in the sea of our other deps

**Comment by Velenir** [comment]
Take a look at now non-experimental #1013

**Pull request closed without merging**